src/metrics-gatherer.ts

```diff
@@ -0,0 +1,230 @@
+import * as prometheus from 'prom-client';
+import * as express from 'express';
+import { TypedError } from 'typed-error'
+
+import * as Debug from 'debug';
+const debug = Debug('node-metrics-gatherer');
+
+import { 
+	LabelSet,
+	DescriptionMap,
+	ExistMap,
+	CustomParams,
+	CustomParamsMap,
+	MetricConstructor,
+	ConstructorMap,
+	MetricsMap,
+	KindMap
+} from './types';
+
+type AuthTestFunc = (req: express.Request) => boolean;
+
+export class MetricsGathererError extends TypedError {}
+    
+export class MetricsGatherer {
+
+	private metrics: MetricsMap
+	private customParams: CustomParamsMap
+	private descriptions: DescriptionMap
+	private existMap: ExistMap
+	private kinds: KindMap
+
+	constructor() {
+		this.initState();
+	}
+
+	private initState() {
+		this.metrics = {
+			gauge: {},
+			counter: {},
+			summary: {},
+			histogram: {},
+		}
+		this.customParams = {};
+		this.descriptions = {};
+		this.existMap = {};
+		this.kinds = {};
+	}
+
+	// fetch the description for a metric
+	describe(name : string, text : string, custom : CustomParams = {}) {
+		if (this.descriptions[name]) {
+			throw new MetricsGathererError(`tried to describe metric "${name}" twice`);
+		}
+		this.descriptions[name] = text;
+		this.customParams[name] = custom;
+	}
+
+	// observe a gauge metric
+	gauge(name : string, 
+		val : number,
+		labels : LabelSet = {}) {
+		this.ensureExists(name, 'gauge', {
+			labelNames: Object.keys(labels)
+		});
+		(<prometheus.Gauge>this.metrics.gauge[name]).set(labels, val);
+	}
+
+	// increment a counter or gauge
+	inc(name : string, 
+		val : number = 1,
+		labels : LabelSet = {}) {
+		// ensure either that this metric already exists, or if not, create a gauge
+		this.ensureExists(name, 'gauge', {
+			labelNames: Object.keys(labels)
+		});
+		if (!this.checkMetricType(name, ['gauge', 'counter'])) {
+			throw new MetricsGathererError(`Tried to increment non-gauge, non-counter metric ${name}`);
+		}
+		if (this.kinds[name] === 'gauge') {
+			(<prometheus.Gauge>this.metrics.gauge[name]).inc(labels, val);
+		} else {
+			(<prometheus.Counter>this.metrics.counter[name]).inc(labels, val);
```

Comment:
It would be better to type `MetricsMap` more strictly and remove the need for this cast